File: sdk/src/test/java/eu/darken/porter/sdk/PorterReadinessTest.java
package eu.darken.porter.sdk;

import static org.junit.Assert.assertEquals;

import android.os.Bundle;

import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import eu.darken.porter.server.IPorterApplication;

/** When a sticky listener catches up on its own and when the dispatch is what calls it. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class PorterReadinessTest {

    private static final String PACKAGE = "eu.darken.porter.probe";

    /** Registers a sticky listener from inside attach, while the connection is neither yet nor no longer. */
    private static class RegisteringService extends FakePorterService {

        final int[] calls = {0};

        @Override
        public Bundle attach(IPorterApplication application, Bundle args) {
            Porter.addBinderReceivedListenerSticky(() -> calls[0]++);
            return super.attach(application, args);
        }
    }

    @After
    public void teardown() {
        Porter.resetForTest();
    }

    @Test
    public void aStickyListenerRegisteredDuringAReplacingAttachIsCalledOnce() {
        Porter.onBinderReceived(new FakePorterService(), PACKAGE);

        RegisteringService replacement = new RegisteringService();
        Porter.onBinderReceived(replacement, PACKAGE);

        assertEquals(1, replacement.calls[0]);
    }

    @Test
    public void aStickyListenerRegisteredDuringTheFirstAttachIsCalledOnce() {
        RegisteringService first = new RegisteringService();

        Porter.onBinderReceived(first, PACKAGE);

        assertEquals(1, first.calls[0]);
    }

    @Test
    public void anAttachThatFailedLeavesNothingToCatchUpOn() {
        FakePorterService refused = new FakePorterService();
        refused.attachFailure = new SecurityException("not an attached client");
        Porter.onBinderReceived(refused, PACKAGE);

        int[] calls = {0};
        Porter.addBinderReceivedListenerSticky(() -> calls[0]++);

        assertEquals(0, calls[0]);
    }

    @Test
    public void aConnectionThatIsGoneLeavesNothingToCatchUpOn() {
        Porter.onBinderReceived(new FakePorterService(), PACKAGE);
        Porter.onBinderReceived(null, null);

        int[] calls = {0};
        Porter.addBinderReceivedListenerSticky(() -> calls[0]++);

        assertEquals(0, calls[0]);
    }
}

File: sdk/src/test/java/eu/darken/porter/sdk/PorterReadinessTest.kt
package eu.darken.porter.sdk

import android.os.Bundle
import eu.darken.porter.server.IPorterApplication
import org.junit.After
import org.junit.Assert.assertNull
import org.junit.Assert.assertSame
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/** What a reader sees in the connection flow while an attach is still running, and after one that went nowhere. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
internal class PorterReadinessTest {

    /** Reads the flow from inside attach, while the connection is neither yet nor no longer. */
    private class ReadingService : FakePorterService() {

        var seenDuringAttach: PorterConnection? = null
        var read = false

        override fun attach(application: IPorterApplication, args: Bundle): Bundle? {
            seenDuringAttach = Porter.connection.value
            read = true
            return super.attach(application, args)
        }
    }

    @After
    fun teardown() {
        Porter.resetForTest()
    }

    @Test
    fun aReaderDuringAReplacingAttachSeesTheConnectionThatIsStillServing() {
        val serving = FakePorterService()
        Porter.onBinderReceived(serving, PACKAGE)
        val servingConnection = Porter.connection.value

        val replacement = ReadingService()
        Porter.onBinderReceived(replacement, PACKAGE)

        assertSame(servingConnection, replacement.seenDuringAttach)
        assertSame(replacement, Porter.connection.value!!.binder)
    }

    @Test
    fun aReaderDuringTheFirstAttachSeesNoConnectionYet() {
        val first = ReadingService()

        Porter.onBinderReceived(first, PACKAGE)

        assertNull(first.seenDuringAttach)
        assertSame(first, Porter.connection.value!!.binder)
    }

    @Test
    fun anAttachThatFailedLeavesNothingToCatchUpOn() {
        val refused = FakePorterService()
        refused.attachFailure = SecurityException("not an attached client")
        Porter.onBinderReceived(refused, PACKAGE)

        assertNull(Porter.connection.value)
    }

    @Test
    fun aConnectionThatIsGoneLeavesNothingToCatchUpOn() {
        Porter.onBinderReceived(FakePorterService(), PACKAGE)
        Porter.onBinderReceived(null, PACKAGE)

        assertNull(Porter.connection.value)
    }

    private companion object {
        const val PACKAGE = "eu.darken.porter.probe"
    }
}
